mesonbuild/interpreter/interpreter.py

```diff
@@ -2079,8 +2080,9 @@ def func_install_headers(self, node: mparser.BaseNode,
             if os.path.isabs(install_subdir):
                 mlog.deprecation('Subdir keyword must not be an absolute path. This will be a hard error in the next release.')
 
-        h = build.Headers(source_files, install_subdir, kwargs['install_dir'],
-                          kwargs['install_mode'], self.subproject)
+        h = build.Headers(source_files, os.path.join(self.environment.source_dir, self.subdir + "/"),
+                install_subdir, kwargs['install_dir'],
+                    kwargs['install_mode'], self.subproject, kwargs['nobase'])
```

Comment:
Instead of moving this parsing logic into the `build.Headers` class, the backend, and minstall itself, what about having the interpreter group these files by leading path and generate a build.Headers per destination subdir?